mirrord-agent/src/steal.rs

```diff
@@ -0,0 +1,374 @@
+use std::{
+    collections::{HashMap, HashSet},
+    net::SocketAddr,
+    path::PathBuf,
+};
+
+use mirrord_protocol::{
+    tcp::{DaemonTcp, LayerStealTcp, NewTcpConnection, TcpClose, TcpData},
+    ConnectionID, Port,
+};
+use rand::distributions::{Alphanumeric, DistString};
+use streammap_ext::StreamMap;
+use tokio::{
+    io::{AsyncWriteExt, ReadHalf, WriteHalf},
+    net::{TcpListener, TcpStream},
+    select,
+    sync::mpsc::{Receiver, Sender},
+};
+use tokio_stream::StreamExt;
+use tokio_util::io::ReaderStream;
+use tracing::{debug, error, info, log::warn};
+
+use crate::{
+    error::{AgentError, Result},
+    runtime::set_namespace,
+};
+
+struct SafeIpTables {
+    inner: iptables::IPTables,
+    chain_name: String,
+}
+
+const IPTABLES_TABLE_NAME: &str = "nat";
+
+impl SafeIpTables {
+    pub fn new() -> Result<Self> {
+        let ipt = iptables::new(false).unwrap();
+        let random_string = Alphanumeric.sample_string(&mut rand::thread_rng(), 5);
+        let chain_name = format!("MIRRORD_REDIRECT_{}", random_string);
+        ipt.new_chain(IPTABLES_TABLE_NAME, &chain_name)
+            .map_err(|e| AgentError::IPTablesError(e.to_string()))?;
+        ipt.append(IPTABLES_TABLE_NAME, &chain_name, "-j RETURN")
+            .map_err(|e| AgentError::IPTablesError(e.to_string()))?;
+        ipt.append(
+            IPTABLES_TABLE_NAME,
+            "PREROUTING",
+            &format!("-j {}", chain_name),
+        )
+        .map_err(|e| AgentError::IPTablesError(e.to_string()))?;
+        Ok(Self {
+            inner: ipt,
+            chain_name,
+        })
+    }
+
+    pub fn add_redirect(&mut self, redirected_port: Port, target_port: Port) -> Result<()> {
+        let rule = format!(
```

Comment:
This is the same `rule` in `remove_redirect`, so maybe extract it and share? I would suggest a helper function that returns the rule `String`.